Src/AutoFixture/Kernel/TypeEnvy.cs

```diff
@@ -11,17 +11,12 @@ internal static class TypeEnvy
         public static TAttribute GetAttribute<TAttribute>(object candidate)
             where TAttribute : Attribute
         {
-            var member = candidate as MemberInfo;
-            if(member != null)
-            {
-                return member.GetCustomAttribute<TAttribute>();
-            }
-            var parameter = candidate as ParameterInfo;
-            if(parameter != null)
-            {
-                return parameter.GetCustomAttribute<TAttribute>();
-            }
-            return null;
+            var attributeProvider = candidate as ICustomAttributeProvider;
+
+            return attributeProvider?
+                .GetCustomAttributes(typeof(TAttribute), true)
+                .Cast<TAttribute>()
```

Comment:
@adamchester Thanks for the reply. Now I better understand your concerns given that you fixed the similar issues in past. 

I've profiled our unit tests and found that this code is pretty hot:
![image](https://user-images.githubusercontent.com/1074182/28776501-9165b414-75ff-11e7-96cf-1f7a458a7714.png)

Therefore, I've refactored it to avoid Linq usage. I'm still not convinced that it worth it, but let's proceed with this premature optimization if we did something similar in past :)